File: leetcode/c++/Edit_Distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size();
        int n = word2.size();
        
        vector<vector<int>> tbl(m+1, vector<int>(n+1, 0));
        
        for(int i = 0; i <= m; i++){
            for(int j = 0; j <= n; j++){
                if(i == 0) tbl[i][j] = j;
                if(j == 0) tbl[i][j] = i;
                
                if(i > 0 && j > 0){
                    if(word1[i-1] == word2[j-1]){
                        tbl[i][j] = tbl[i-1][j-1];
                    }else{
                        tbl[i][j] = min(min(tbl[i-1][j-1], tbl[i-1][j]), tbl[i][j-1])+1;
                    }
                }
            }
        }
        
        return tbl[m][n];
    }
};

File: leetcode/java/EditDistance.java
class Solution {
    public int minDistance(String w1, String w2) {
        if(w1 == null || w2 == null) throw new IllegalArgumentException("xx");
        int m = w1.length();
        int n = w2.length();
        
        int[][] t = new int[m+1][n+1];
        for(int i = 0; i <= m; i++) {
            for(int j = 0; j <= n; j++) {
                if(i == 0) t[i][j] = j;
                else if(j == 0) t[i][j] = i;
                else {
                    if(w1.charAt(i-1) == w2.charAt(j-1)) t[i][j] = t[i-1][j-1];
                    else {
                        t[i][j] = Math.min(Math.min(t[i-1][j], t[i][j-1]), t[i-1][j-1]) +1;
                    }
                }
            }
        }
        return t[m][n];
    }
}
